stdlib/compileall.pyi

```diff
@@ -1,17 +1,20 @@
 import sys
-from _typeshed import AnyPath
-from typing import Any, Optional, Pattern
+from _typeshed import AnyPath, _T_contra
+from typing import Any, AnyStr, Optional, Protocol
 
 if sys.version_info >= (3, 7):
     from py_compile import PycInvalidationMode
 
+class _SupportsSearch(Protocol[_T_contra]):
+    def search(self, string: _T_contra) -> Any: ...
+
 if sys.version_info >= (3, 9):
     def compile_dir(
         dir: AnyPath,
         maxlevels: Optional[int] = ...,
         ddir: Optional[AnyPath] = ...,
         force: bool = ...,
-        rx: Optional[Pattern[Any]] = ...,
+        rx: Optional[_SupportsSearch[AnyStr]] = ...,
```

Comment:
The AnyStr doesn't make sense to me here; how is the typevar going to get constrained? I'm guessing it depends on whether `dir` is a bytes or a str path, so perhaps that argument should also have a TypeVar in it.

(Note that AnyPath and AnyStr are very different things: the first is a union and the second a TypeVar.)

Reply:
OK, so compile_dir runs compile_file on every file it can using listdir - which means file will be bytes or str, so rx there can accept str and bytes. which is why I used AnyStr, a type that can be either or.

compile_file however gets AnyPath as file name and can pass that along to rx as is if quiet is 2. Meaning that the rx in compile_file might need to handle the same type as fullname. Which means I might need to fix that? Not sure what to do there.

Either way, I'm not following you. What did you have in mind?

Reply:
Well, the place I started with is that having a single TypeVar in a signature tends to be a mistake, because the point of a typevar is to use the same type in multiple places. See the discussion in https://mail.python.org/archives/list/typing-sig@python.org/message/NRFNHGXHXPGBR6FP3TIOZZ6VS4XJZX6K/ for some context.

But I just tried it out and at least in 3.7, we don't need all these Unions: compile_dir actually only works on str paths, not bytes paths. In `compile_file`, it calls `importlib.util.cache_from_source`, and that only accepts str, not bytes. So instead, let's just change all the parameters dealing with paths to accept only strs.